File: platform/ios/overlay/submodules/NebulaLinkUI/Sources/NebulaLatency.swift
import Foundation

// Presentation only: raw GET timing and cache values remain unchanged.
enum NebulaLatency {
    static func isMeasured(_ rawMs: Int, checkedAt: Int64) -> Bool {
        rawMs > 0 || (rawMs == 0 && checkedAt > 0)
    }

    static func format(_ rawMs: Int, method: String, checkedAt: Int64,
                       unit: String, unknown: String, failed: String) -> String {
        if rawMs < 0 { return failed }
        if !isMeasured(rawMs, checkedAt: checkedAt) { return unknown }
        if method == "nimbo" { return "≈\(Int((Double(rawMs) / 3.3).rounded())) \(unit)" }
        return "\(rawMs) \(unit)"
    }
}

File: platform/android/overlay/TMessagesProj/src/main/java/app/nebulagram/ui/NebulaLatency.java
package app.nebulagram.ui;

/** Presentation only. Never write these estimates back to the core or its cache. */
public final class NebulaLatency {
    private NebulaLatency() {}

    public static boolean isMeasured(int rawMs, long checkedAt) {
        return rawMs > 0 || (rawMs == 0 && checkedAt > 0);
    }

    public static long displayMillis(int rawMs, String method) {
        return rawMs >= 0 && "nimbo".equals(method) ? Math.round(rawMs / 3.3) : rawMs;
    }

    public static String format(int rawMs, String method, long checkedAt,
                                String unit, String unknown, String failed) {
        if (rawMs < 0) return failed;
        if (!isMeasured(rawMs, checkedAt)) return unknown;
        return ("nimbo".equals(method) ? "≈" : "") + displayMillis(rawMs, method) + " " + unit;
    }
}
